packages/integration-tests/src/retry.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Executes an asynchronous operation with retry logic.
+ * This function attempts to execute the provided callable function multiple times
+ * based on the specified retry conditions. It's useful for handling transient
+ * errors or temporary service unavailability.
+ */
+export const runWithRetry = async <T>(
+  callable: () => Promise<T>,
+  retryPredicate: (error: Error) => boolean,
+  maxAttempts = 3,
+  timeout?: number
+): Promise<T> => {
+  const startTime = Date.now();
+  const collectedErrors: Error[] = [];
+
+  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
+    try {
+      if (timeout && Date.now() - startTime >= timeout) {
+        throw new Error(`Operation timed out after ${timeout}ms`);
+      }
+
+      const result = await callable();
+      return result;
+    } catch (error) {
+      if (error instanceof Error) {
+        collectedErrors.push(error);
+
+        if (!retryPredicate(error) || attempt === maxAttempts) {
+          break;
+        }
+
+        console.log(`Attempt ${attempt} failed. Retrying...`);
+        await new Promise((resolve) => setTimeout(resolve, 5000)); // 5 second delay between retries
```

Comment:
I'd remove delay for now.

we don't seem to have data that would say it's really needed. we should add this later if we see that operations still fail under retry and that errors are related to throttling.